### PR TITLE
test: error for different shape on array Assignment

### PR DESCRIPTION
**Pull request by jinangshah21**
Done these things :
1) Improved Error Message for incompatible ranks.
2) Throw Error if different shape on array assignment.

### CODE REVIEW

**Comment by Pranavchiku** [comment]
Okay so you just need to rebase this PR, follow the steps show below to clean git history ( assuming you are in temp_branch )

```
git checkout -b temp_branch_2
git checkout main
git pull origin main
git checkout temp_branch
git rebase main
git push <your_remote> temp_branch -f
```

**Review by Pranavchiku — CHANGES_REQUESTED** [review]
This PR looks good to me, clean commit history and ping if any problem occurs.

**Comment by gxyd** [comment]
I moved it to "draft", as Pranav left some comments which need to be addressed.

File: src/lfortran/semantics/ast_body_visitor.cpp [hunk 1]
```diff
@@ -2328,8 +2328,7 @@ class BodyVisitor : public CommonVisitor<BodyVisitor> {
         // real :: x(4); x = [1, 2, 3, 4] to be a valid assignment (as RHS is "integer array")
         // TODO: the only reason to not do this check for "reshape" is because
         // incorrect 'n_dims' and 'shape' returned for "reshape" currently
-        if (target_type->type == ASR::ttypeType::Array && value_type->type == ASR::ttypeType::Array
-            && value->type != ASR::exprType::ArrayReshape) {
+        if (target_type->type == ASR::ttypeType::Array && value_type->type == ASR::ttypeType::Array) {
```
Comment (Pranavchiku) [inline src/lfortran/semantics/ast_body_visitor.cpp:2331]:
```suggestion
        if (target_type != nullptr && value_type != nullptr && ASRUtils::is_array(target_type) && ASRUtils::is_array(value_type)) {
```

File: src/lfortran/semantics/ast_common_visitor.h [hunk 2]
```diff
@@ -3007,9 +3007,29 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
                         value = init_expr;
                     }
                     ASR::ttype_t *init_type = ASRUtils::expr_type(init_expr);
-                    if(ASRUtils::extract_n_dims_from_ttype(init_type)!=ASRUtils::extract_n_dims_from_ttype(type) && ASRUtils::extract_n_dims_from_ttype(type)==0){
-                        throw SemanticError("Initialisation using ArrayConstant is supported only for single dimensional arrays",x.base.base.loc);
+
+                    size_t rhs_rank = ASRUtils::extract_n_dims_from_ttype(init_type);
+                    size_t lhs_rank = ASRUtils::extract_n_dims_from_ttype(type);
+
+                    if(lhs_rank!=rhs_rank){
```
Comment (Pranavchiku) [inline src/lfortran/semantics/ast_common_visitor.h:3014]:
```suggestion
                    if ( lhs_rank != rhs_rank ) {
```

**Review by Pranavchiku — CHANGES_REQUESTED** [review]
apply the following changes, build and update references

File: tests/reference/asr-arrays_reshape_14-913f34e.stderr [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    void visit_unit_decl2(const unit_decl2_t &b) { visit_unit_decl2_t(b, self()); }
+[2m  File "[0m[1m[35m$DIR/src/lfortran/ast.h[39m[0m[2m", line 4500, in void LCompilers::LFortran::AST::visit_unit_decl2_t<LCompilers::LFortran::SymbolTableVisitor>(LCompilers::LFortran::AST::unit_decl2_t const&, LCompilers::LFortran::SymbolTableVisitor&)[0m
+    case unit_decl2Type::Declaration: { v.visit_Declaration((const Declaration_t &)x); return; }
+[2m  File "[0m[1m[35m$DIR/src/lfortran/semantics/ast_symboltable_visitor.cpp[39m[0m[2m", line 1566, in LCompilers::LFortran::SymbolTableVisitor::visit_Declaration(LCompilers::LFortran::AST::Declaration_t const&)[0m
+    visit_DeclarationUtil(x);
+[2m  File "[0m[1m[35m$DIR/src/lfortran/semantics/ast_common_visitor.h[39m[0m[2m", line 2932, in LCompilers::LFortran::CommonVisitor<LCompilers::LFortran::SymbolTableVisitor>::visit_DeclarationUtil(LCompilers::LFortran::AST::Declaration_t const&)[0m
+    if(value->type==63 && type->type==0){
+[2m  Binary file "[0m[1m[35m/lib/x86_64-linux-gnu/libc.so.6[39m[0m[2m", in <invalid>[0m
```
Comment (Pranavchiku) [inline tests/reference/asr-arrays_reshape_14-913f34e.stderr:38]: this should not be updated, with the suggested changes this shall go.

**Review by Pranavchiku — APPROVED** [review]
this looks good to me. I'll wait for one more review before merging.

**Comment by Pranavchiku** [comment]
A good practice is to start using conventional commits, you may read more about it from https://www.conventionalcommits.org/en/v1.0.0/. Also, make sure you resolve conversation after addressing the changes.

File: src/lfortran/semantics/ast_common_visitor.h [hunk 4]
```diff
@@ -3006,9 +3006,29 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
...
+
+                     if ( ASR::is_a<ASR::Array_t>(*init_type) && !ASR::is_a<ASR::ImpliedDoLoop_t>(*init_expr)){
+                        ASR::Array_t* arr_rhs = ASR::down_cast<ASR::Array_t>(init_type);
+                        ASR::Array_t* arr_lhs = ASR::down_cast<ASR::Array_t>(type);
+                        
+                    for (size_t i = 0; i < arr_lhs->n_dims; i++) {
+                           std::string lhs_dim = ASRUtils::extract_dim_value(arr_lhs->m_dims[i].m_length);
+                           std::string rhs_dim = ASRUtils::extract_dim_value(arr_rhs->m_dims[i].m_length);
+                            if(lhs_dim!=":" && rhs_dim!=":" && lhs_dim!=rhs_dim){
+                                    throw SemanticError("Incompatible shape of array on assignment on dimension " + std::to_string(i) +
```
Comment (certik) [inline src/lfortran/semantics/ast_common_visitor.h:3027]: We should count from 1, not from 0:
```suggestion
                                    throw SemanticError("Incompatible shape of array on assignment on dimension " + std::to_string(i+1) +
```

**Review by certik — APPROVED** [review]
I think this looks great! Thanks for fixing it. I left a minor comment that can be addressed in a subsequent PR.

**Pull request merged**